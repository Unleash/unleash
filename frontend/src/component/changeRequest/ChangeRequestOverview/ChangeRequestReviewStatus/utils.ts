export const getBrowserTimezoneInHumanReadableUTCOffset = (
    date = new Date(),
): string => {
    const offset = -date.getTimezoneOffset();
    const hours = Math.floor(Math.abs(offset) / 60);
    const minutes = Math.abs(offset) % 60;
    let sign = '+';
    if (offset < 0) {
        sign = '-';
    }

    // Ensure that hours and minutes are two digits
    const zeroPaddedHours = hours.toString().padStart(2, '0');
    const zeroPaddedMinutes = minutes.toString().padStart(2, '0');

    return `UTC${sign}${zeroPaddedHours}:${zeroPaddedMinutes}`;
};

export const getBrowserTimezone = (): string => {
    return Intl.DateTimeFormat().resolvedOptions().timeZone;
};
