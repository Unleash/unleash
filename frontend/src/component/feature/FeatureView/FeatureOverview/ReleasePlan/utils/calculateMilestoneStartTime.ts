import type { IReleasePlanMilestone } from 'interfaces/releasePlans';
import { addMinutes } from 'date-fns';

const parseStartTime = (startedAt: string | null | undefined): Date | null => {
    if (!startedAt) return null;

    const date = new Date(startedAt);
    if (Number.isNaN(date.getTime())) {
        return null;
    }
    return date;
};

const getIntervalMinutes = (
    milestone: IReleasePlanMilestone,
): number | null => {
    const intervalMinutes = milestone.transitionCondition?.intervalMinutes;

    if (!intervalMinutes) {
        return null;
    }
    return intervalMinutes;
};

const findMostRecentStartedMilestone = (
    milestones: IReleasePlanMilestone[],
    targetIndex: number,
): { index: number; startTime: Date } | null => {
    for (let i = targetIndex; i >= 0; i--) {
        const startTime = parseStartTime(milestones[i].startedAt);
        if (startTime) {
            return { index: i, startTime };
        }
    }
    return null;
};

const findBaselineMilestone = (
    milestones: IReleasePlanMilestone[],
    targetIndex: number,
    activeMilestoneId?: string,
): { index: number; startTime: Date } | null => {
    if (!activeMilestoneId) {
        return findMostRecentStartedMilestone(milestones, targetIndex);
    }

    const activeIndex = milestones.findIndex((m) => m.id === activeMilestoneId);
    if (activeIndex === -1 || activeIndex > targetIndex) {
        return findMostRecentStartedMilestone(milestones, targetIndex);
    }

    const activeStartTime = parseStartTime(milestones[activeIndex].startedAt);
    if (activeStartTime) {
        return { index: activeIndex, startTime: activeStartTime };
    }

    return findMostRecentStartedMilestone(milestones, targetIndex);
};

const calculateTimeFromBaseline = (
    milestones: IReleasePlanMilestone[],
    baseline: { index: number; startTime: Date },
    targetIndex: number,
): Date | null => {
    let currentTime = baseline.startTime;

    for (let i = baseline.index; i < targetIndex; i++) {
        const previousMilestone = milestones[i];
        const intervalMinutes = getIntervalMinutes(previousMilestone);

        if (!intervalMinutes) return null;

        currentTime = addMinutes(currentTime, intervalMinutes);
    }

    return currentTime;
};

export const calculateMilestoneStartTime = (
    milestones: IReleasePlanMilestone[],
    targetMilestoneId: string,
    activeMilestoneId?: string,
): Date | null => {
    const targetIndex = milestones.findIndex((m) => m.id === targetMilestoneId);
    if (targetIndex === -1) return null;

    const baseline = findBaselineMilestone(
        milestones,
        targetIndex,
        activeMilestoneId,
    );
    if (!baseline) return null;

    if (baseline.index === targetIndex) {
        return baseline.startTime;
    }

    return calculateTimeFromBaseline(milestones, baseline, targetIndex);
};
