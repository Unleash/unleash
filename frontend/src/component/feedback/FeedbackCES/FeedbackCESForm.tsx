import { useStyles } from 'component/feedback/FeedbackCES/FeedbackCESForm.styles';
import { Button, TextField } from '@mui/material';
import React, { useState } from 'react';
import produce from 'immer';
import useToast from 'hooks/useToast';
import { IFeedbackCESState } from 'component/feedback/FeedbackCESContext/FeedbackCESContext';
import { FeedbackCESScore } from 'component/feedback/FeedbackCES/FeedbackCESScore';
import { sendFeedbackInput } from 'component/feedback/FeedbackCES/sendFeedbackInput';

export interface IFeedbackCESFormProps {
    state: IFeedbackCESState;
    onClose: () => void;
}

export interface IFeedbackCESForm {
    score: number;
    comment: string;
    path: string;
}

export const FeedbackCESForm = ({ state, onClose }: IFeedbackCESFormProps) => {
    const [loading, setLoading] = useState(false);
    const { setToastData } = useToast();
    const { classes: styles } = useStyles();

    const [form, setForm] = useState<Partial<IFeedbackCESForm>>({
        path: state.path,
    });

    const onCommentChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
        setForm(
            produce(draft => {
                draft.comment = event.target.value;
            })
        );
    };

    const onSubmit = async (event: React.FormEvent) => {
        event.preventDefault();

        if (loading) {
            return;
        }

        try {
            setLoading(true);
            await sendFeedbackInput(form);
            setToastData({
                type: 'success',
                title: 'Feedback sent. Thank you!',
                confetti: true,
            });
            onClose();
        } finally {
            setLoading(false);
        }
    };

    return (
        <div className={styles.container}>
            <h1 className={styles.title}>Please help us improve</h1>
            <form
                className={styles.form}
                onSubmit={onSubmit}
                aria-live="polite"
            >
                <p className={styles.subtitle}>{state.title}</p>
                <FeedbackCESScore form={form} setForm={setForm} />
                <div hidden={!form.score}>
                    <label htmlFor="comment" className={styles.textLabel}>
                        {state.text}
                    </label>
                    <TextField
                        value={form.comment ?? ''}
                        onChange={onCommentChange}
                        multiline
                        rows={3}
                        variant="outlined"
                        fullWidth
                    />
                </div>
                <div className={styles.buttons} hidden={!form.score}>
                    <Button
                        type="submit"
                        color="primary"
                        variant="contained"
                        className={styles.button}
                        disabled={!form.score || loading}
                    >
                        Send feedback
                    </Button>
                </div>
            </form>
        </div>
    );
};
