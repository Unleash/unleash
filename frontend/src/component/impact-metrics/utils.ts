export const getTimeUnit = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
            return 'minute';
        case 'day':
            return 'hour';
        case 'week':
            return 'day';
        case 'month':
            return 'week';
        default:
            return 'day';
    }
};

export const getDisplayFormat = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
            return 'HH:mm';
        case 'day':
            return 'HH:mm';
        case 'week':
            return 'MMM dd';
        case 'month':
            return 'MMM dd';
        default:
            return 'MMM dd';
    }
};

export const getSeriesLabel = (metric: Record<string, string>): string => {
    const { __name__, ...labels } = metric;

    const labelParts = Object.entries(labels)
        .filter(([key, value]) => key !== '__name__' && value)
        .map(([key, value]) => `${key}=${value}`)
        .join(', ');

    if (!__name__ && !labelParts) {
        return 'Series';
    }

    if (!__name__) {
        return labelParts;
    }

    if (!labelParts) {
        return __name__;
    }

    return `${__name__} (${labelParts})`;
};

export const formatLargeNumbers = (value: number): string => {
    if (value >= 1000000) {
        return `${(value / 1000000).toFixed(0)}M`;
    }
    if (value >= 1000) {
        return `${(value / 1000).toFixed(0)}k`;
    }
    return value.toString();
};

export const getMetricType = (seriesName: string) => {
    if (seriesName.startsWith('unleash_counter_')) return 'counter';
    if (seriesName.startsWith('unleash_gauge_')) return 'gauge';
    return 'unknown';
};
