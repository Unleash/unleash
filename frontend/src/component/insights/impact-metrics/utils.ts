export const getTimeUnit = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
            return 'minute';
        case 'day':
            return 'hour';
        case 'week':
            return 'day';
        case 'month':
            return 'day';
        default:
            return 'hour';
    }
};

export const getDisplayFormat = (selectedRange: string) => {
    switch (selectedRange) {
        case 'hour':
        case 'day':
            return 'HH:mm';
        case 'week':
        case 'month':
            return 'MMM dd';
        default:
            return 'MMM dd HH:mm';
    }
};

export const getSeriesLabel = (metric: Record<string, string>): string => {
    const { __name__, ...labels } = metric;

    const labelParts = Object.entries(labels)
        .filter(([key, value]) => key !== '__name__' && value)
        .map(([key, value]) => `${key}=${value}`)
        .join(', ');

    if (!__name__ && !labelParts) {
        return 'Series';
    }

    if (!__name__) {
        return labelParts;
    }

    if (!labelParts) {
        return __name__;
    }

    return `${__name__} (${labelParts})`;
};

export const formatLargeNumbers = (value: number): string => {
    if (value >= 1000000) {
        return `${(value / 1000000).toFixed(0)}M`;
    }
    if (value >= 1000) {
        return `${(value / 1000).toFixed(0)}k`;
    }
    return value.toString();
};
