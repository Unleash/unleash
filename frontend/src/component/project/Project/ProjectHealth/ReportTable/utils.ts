import differenceInDays from 'date-fns/differenceInDays';
import { EXPERIMENT, OPERATIONAL, RELEASE } from 'constants/featureToggleTypes';

const FORTY_DAYS = 40;
const SEVEN_DAYS = 7;

export const toggleExpiryByTypeMap: Record<string, number> = {
    [EXPERIMENT]: FORTY_DAYS,
    [RELEASE]: FORTY_DAYS,
    [OPERATIONAL]: SEVEN_DAYS,
};

export const getDiffInDays = (date: Date, now: Date) => {
    return Math.abs(differenceInDays(date, now));
};

export const expired = (diff: number, type: string) => {
    return diff >= toggleExpiryByTypeMap[type];
};
