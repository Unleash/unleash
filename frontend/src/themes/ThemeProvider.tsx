import React, { FC } from 'react';
import { CssBaseline, ThemeProvider as MuiThemeProvider } from '@mui/material';

import createCache from '@emotion/cache';
import { CacheProvider } from '@emotion/react';
import { useThemeMode } from 'hooks/useThemeMode';

export const muiCache = createCache({
    key: 'mui',
    prepend: true,
});

export const ThemeProvider: FC = ({ children }) => {
    const { resolveTheme } = useThemeMode();

    return (
        <CacheProvider value={muiCache}>
            <MuiThemeProvider theme={resolveTheme()}>
                <CssBaseline />
                {children}
            </MuiThemeProvider>
        </CacheProvider>
    );
};
