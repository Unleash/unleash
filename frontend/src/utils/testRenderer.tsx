import React, { FC } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { SWRConfig } from 'swr';
import { ThemeProvider } from 'themes/ThemeProvider';
import { IPermission } from 'interfaces/user';
import { AnnouncerProvider } from 'component/common/Announcer/AnnouncerProvider/AnnouncerProvider';
import { AccessProviderMock } from 'component/providers/AccessProvider/AccessProviderMock';

export const render = (
    ui: JSX.Element,
    {
        route = '/',
        permissions = [],
        ...renderOptions
    }: { route?: string; permissions?: IPermission[] } & Omit<
        RenderOptions,
        'queries'
    > = {}
) => {
    window.history.pushState({}, 'Test page', route);

    const Wrapper: FC = ({ children }) => (
        <SWRConfig value={{ provider: () => new Map() }}>
            <AccessProviderMock permissions={permissions}>
                <ThemeProvider>
                    <AnnouncerProvider>
                        <BrowserRouter>{children}</BrowserRouter>
                    </AnnouncerProvider>
                </ThemeProvider>
            </AccessProviderMock>
        </SWRConfig>
    );

    return rtlRender(ui, {
        wrapper: Wrapper,
        ...renderOptions,
    });
};
