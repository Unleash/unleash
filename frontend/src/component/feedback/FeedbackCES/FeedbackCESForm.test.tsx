import { FeedbackCESForm } from './FeedbackCESForm';
import { ThemeProvider } from 'themes/ThemeProvider';
import { render } from 'utils/testRenderer';

test('FeedbackCESForm', () => {
    const onClose = () => {
        throw new Error('Unexpected onClose call.');
    };

    render(
        <ThemeProvider>
            <FeedbackCESForm
                onClose={onClose}
                state={{ title: 'a', text: 'b', path: '/c' }}
            />
        </ThemeProvider>
    );

    expect(document.body).toMatchSnapshot();
});
