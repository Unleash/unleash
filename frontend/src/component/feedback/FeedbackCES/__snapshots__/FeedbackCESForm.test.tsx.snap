// Vitest Snapshot v1

exports[`FeedbackCESForm 1`] = `
<body>
  <div>
    <div
      class="css-fdcp7c-container"
    >
      <h1
        class="css-iyd7t0-title"
      >
        Please help us improve
      </h1>
      <form
        aria-live="polite"
        class="css-1rgjoiw-form"
      >
        <p
          class="css-mhztvj-subtitle"
        >
          a
        </p>
        <div
          class="css-io6e1g-scoreInput"
        >
          <span
            class="css-16omcck-scoreHelp"
          >
            Very difficult
          </span>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="1"
            />
            <span>
              1
            </span>
          </label>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="2"
            />
            <span>
              2
            </span>
          </label>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="3"
            />
            <span>
              3
            </span>
          </label>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="4"
            />
            <span>
              4
            </span>
          </label>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="5"
            />
            <span>
              5
            </span>
          </label>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="6"
            />
            <span>
              6
            </span>
          </label>
          <label
            class="css-pq8zjr-scoreValue"
          >
            <input
              name="score"
              type="radio"
              value="7"
            />
            <span>
              7
            </span>
          </label>
          <span
            class="css-16omcck-scoreHelp"
          >
            Very easy
          </span>
        </div>
        <div
          hidden=""
        >
          <label
            class="css-13obo9p-textLabel"
            for="comment"
          >
            b
          </label>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-1qvtpcv-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1wvz5kg-MuiInputBase-input-MuiOutlinedInput-input"
                id="mui-1"
                rows="3"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1wvz5kg-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-hdw1oc"
                >
                  <span
                    class="notranslate"
                  >
                    â€‹
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="css-n2ggaa-buttons"
          hidden=""
        >
          <button
            class="MuiButtonBase-root Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-9rpict-MuiButtonBase-root-MuiButton-root-button"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            Send feedback
          </button>
        </div>
      </form>
    </div>
    <div
      aria-atomic="true"
      aria-live="polite"
      class="css-i8rqz1-container"
      data-testid="ANNOUNCER_ELEMENT_TEST_ID"
      role="status"
    />
  </div>
</body>
`;
